src/compiler/checker.ts

```diff
@@ -5998,20 +5998,39 @@ namespace ts {
                 if (type.flags & TypeFlags.Any) {
                     return type;
                 }
+
                 let signature = getResolvedSignature(expr);
 
                 if (signature.typePredicate &&
-                    expr.arguments[signature.typePredicate.parameterIndex] &&
-                    getSymbolAtLocation(expr.arguments[signature.typePredicate.parameterIndex]) === symbol) {
+                    expr.arguments[signature.typePredicate.parameterIndex]) {
+
+                    let arg = expr.arguments[signature.typePredicate.parameterIndex];
+                    if (arg.kind === SyntaxKind.Identifier && getSymbolAtLocation(arg) === symbol) {
+                        return narrowType();
+                    }
+                    else if (arg.kind === SyntaxKind.AsExpression || arg.kind === SyntaxKind.TypeAssertionExpression) {
+                        if ((<AssertionExpression>arg).expression) {
+                            if ((<AssertionExpression>arg).expression.kind === SyntaxKind.Identifier &&
+                                getSymbolAtLocation((<AssertionExpression>arg).expression) === symbol) {

```

Comment:
This is the same check as above, it's just doing the check on a different expression. I would instead structure it like this:
```ts
let arg = expr.arguments[signature.typePredicate.parameterIndex];
if (arg.kind === SyntaxKind.AsExpression || arg.kind === SyntaxKind.TypeAssertionExpression) {
    arg = (<AssertionExpression>arg).expression;
}
if (arg.kind === SyntaxKind.Identifier && getSymbolAtLocation(arg) === symbol) {
    if (!assumeTrue) ...
}
```
That way you do not need the narrowType inner function